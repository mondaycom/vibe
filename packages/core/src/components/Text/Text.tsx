import React, { forwardRef, ReactNode } from "react";
import cx from "classnames";
import { camelCase } from "lodash-es";
import { getStyle } from "../../helpers/typesciptCssModulesHelper";
import VibeComponent from "../../types/VibeComponent";
import { TextType as TextTypeEnum, TextWeight as TextWeightEnum } from "./TextConstants";
import Typography, { TypographyProps } from "../Typography/Typography";
import {
  TypographyAlign as TypographyAlignEnum,
  TypographyColor as TypographyColorEnum
} from "../Typography/TypographyConstants";
import { TextType, TextWeight } from "./Text.types";
import { withStaticProps } from "../../types";
import styles from "./Text.module.scss";

export interface TextProps extends TypographyProps {
  /**
   * The text style variant.
   */
  type?: TextType;
  /**
   * The font weight of the text.
   */
  weight?: TextWeight;
  /**
   * The content inside the text component.
   */
  children: ReactNode;
}

const Text: VibeComponent<TextProps, HTMLElement> & {
  types?: typeof TextTypeEnum;
  weights?: typeof TextWeightEnum;
  colors?: typeof TypographyColorEnum;
  align?: typeof TypographyAlignEnum;
} = forwardRef(
  (
    {
      className,
      type = "text2",
      weight = "normal",
      ellipsis,
      element = "div",
      children,
      ...typographyProps
    }: TextProps,
    ref
  ) => {
    const overrideEllipsis = ellipsis ?? element !== "p";
    return (
      <Typography
        ref={ref}
        className={cx(styles.text, getStyle(styles, camelCase(type + "-" + weight)), className)}
        ellipsis={overrideEllipsis}
        element={element}
        {...typographyProps}
      >
        {children}
      </Typography>
    );
  }
);

export default withStaticProps(Text, {
  types: TextTypeEnum,
  weights: TextWeightEnum,
  colors: TypographyColorEnum,
  align: TypographyAlignEnum
});
