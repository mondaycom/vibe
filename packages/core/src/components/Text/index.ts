export { default as Text, type TextProps } from "./Text";

export * from "./Text.types";
