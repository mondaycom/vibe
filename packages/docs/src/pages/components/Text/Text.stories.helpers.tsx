import React from "react";
import { StorybookLink, Tip } from "vibe-storybook-components";

export const TipHeading = () => (
  <Tip>
    Check out our{" "}
    <StorybookLink page="Text/Heading" size={StorybookLink.sizes.SMALL}>
      Heading
    </StorybookLink>{" "}
    component for text headlines.
  </Tip>
);

export const TipLink = () => (
  <Tip>
    If you need to place a link outside of the textual flow, please use our{" "}
    <StorybookLink page="Components/Link" size={StorybookLink.sizes.SMALL}>
      Link
    </StorybookLink>{" "}
    component.
  </Tip>
);
