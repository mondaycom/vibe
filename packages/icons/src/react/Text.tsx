/* eslint-disable */
/* tslint:disable */
import * as React from 'react';
export interface TextProps extends React.SVGAttributes<SVGElement> {
size?: string | number;
}
const Text: React.FC<TextProps> = ({size, ...props}) => (
  <svg viewBox="0 0 20 20" fill="currentColor" width={ size || "20" } height={ size || "20" } {...props}>
    <path d="M11.9016 12.6511H8.09212L7.23653 15H6L9.47229 6H10.5215L14 15H12.7697L11.9016 12.6511ZM8.45433 11.6745H11.5457L9.99688 7.46497L8.45433 11.6745Z" fill="currentColor"
    />
  </svg>
);
Text.displayName = 'Text';
export default Text;
/* tslint:enable */
/* eslint-enable */
