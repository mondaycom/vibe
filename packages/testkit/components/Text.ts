import { Page, Locator, test } from "@playwright/test";
import { BaseElement } from "./BaseElement";

/**
 * Class representing a Text element.
 * Extends the BaseElement class.
 */
export class Text extends BaseElement {
  /**
   * Create a Text element.
   * @param {Page} page - The Playwright page object.
   * @param {Locator} locator - The locator for the Text element.
   * @param {string} elementReportName - The name for reporting purposes.
   */
  constructor(page: Page, locator: Locator, elementReportName: string) {
    super(page, locator, elementReportName);
  }

  /**
   * Set the text of the Text element.
   * @param {string} text - The text to set.
   * @returns {Promise<void>} A promise that resolves when the text is set.
   */
  async setText(text: string): Promise<void> {
    return await test.step(`Set text for ${this.getElementReportName()}`, async () => {
      await this.getLocator().fill(text);
    });
  }
}
