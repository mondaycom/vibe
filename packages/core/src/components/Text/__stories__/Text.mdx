import { Meta } from "@storybook/blocks";
import { Tip } from "vibe-storybook-components";
import Text from "../Text";
import Flex from "../../Flex/Flex";
import Button from "../../Button/Button";
import Icon from "../../Icon/Icon";
import { Link } from "@vibe/icons";
import {
  EDITABLE_HEADING,
  HEADING,
  LINK
} from "../../../storybook/components/related-components/component-description-map";
import { TipHeading, TipLink } from "./Text.stories.helpers";
import * as TextStories from "./Text.stories";
import styles from "./Text.stories.module.scss";

<Meta of={TextStories} />

# Text

- [Overview](#overview)
- [Props](#props)
- [Usage](#usage)
- [Variants](#variants)
- [Do’s and don’ts](#dos-and-donts)
- [Related components](#related-components)
- [Feedback](#feedback)

## Overview

The text component serves as a wrapper for applying typography styles to the text it contains.

<Canvas of={TextStories.Overview} />

### Import

```js
import { Text } from "@vibe/core";
```

<TipHeading />

## Props

<PropsTable />

## Usage

<UsageGuidelines guidelines={["Use it to display text formed from a single sentence, or multiple sentences."]} />
<UsageGuidelines
  guidelines={[
    "Use Text3 for short caption text, label or as explanation text in menus. In any other case, use Text1 or Text 2."
  ]}
/>

## Variants

### Sizes and weights

Text component comes in three sizes: text1 (16px), text2 (14px), text3 (12px) , and in three weights: bold (700), medium (600) and normal (400)

<Canvas of={TextStories.SizesAndWeights} />

### Colors

Text component comes in four colors: primary, secondary, on-primary and on-inverted

<Canvas of={TextStories.Colors} />

### Overflow

Our Text component supports overflow state.
When the text is longer than its container and the ellipsis flag is on, the end of the text will be truncated and will display "..."

We support two kinds of ellipsis: single-line ellipsis with a tooltip displayed in hover or ellipsis after multiple lines. You can see examples for both use cases below.

<Canvas of={TextStories.Overflow} />

<Tip>Ellipsis prop is true by default. If you wish to turn off ellipsis you can change the prop to false.</Tip>

### Paragraph

To use the Text component for a paragraph, utilize the element prop with value of `"p"`.
This changes the text wrapper element to `p`, enabling the creation of paragraph-style text without ellipsis by default.
The paragraph will receive default top and bottom margins based on browser settings for `p` elements.
Customize ellipsis behavior using the "ellipsis" prop or override default margins with a custom class name.

<Canvas of={TextStories.Paragraph} />

### Links

A Text component with a link skin can be used to create a link within running text that redirects to an external webpage, as demonstrated in the following example.

<Canvas of={TextStories.LinksInsideRunningText} />

<TipLink />

## Do’s and Don’ts

<ComponentRules
  rules={[
    {
      positive: {
        component: (
          <Text type="text1" element="span">
            The quick brown fox jumps over the{" "}
            <Text type="text1" weight="bold" element="span">
              lazy dog
            </Text>
          </Text>
        ),
        description: "You can combine two font weights in one sentence to create an emphasis."
      },
      negative: {
        component: (
          <Text type="text2" element="span">
            The quick brown fox jumps over the{" "}
            <Text type="text1" element="span">
              lazy dog
            </Text>
          </Text>
        ),
        description: "Don't use more then one font size in a sentence."
      }
    },
    {
      positive: {
        component: (
          <Flex style={{ width: "70%" }} direction="column" justify="center" align="center" gap="small">
            <Text maxLines={3}>
              monday CRM lets you control your entire sales funnel and close more deals, faster. Automate manual work
              and streamline sales activities from start to finish. Explore monday sales CRM templates
            </Text>
            <Button kind="tertiary">Read more</Button>
          </Flex>
        ),
        description: "If ellipses are used in a paragraph, always use a CTA to reveal more information."
      },
      negative: {
        component: (
          <Text maxLines={3} style={{ width: "70%" }}>
            monday CRM lets you control your entire sales funnel and close more deals, faster. Automate manual work and
            streamline sales activities from start to finish. Explore monday sales CRM templates
          </Text>
        ),
        description: "Don't use ellipsis without providing a way for the user to read the full text."
      }
    },
    {
      positive: {
        component: (
          <Flex direction="column" align="start" gap="xs" className={styles.textBoxContainer}>
            <Flex style={{ width: "90%" }} gap="small">
              <Icon icon={Link} />
              <Text color="secondary" maxLines={3} type="text3">
                Section title in text3
              </Text>
            </Flex>
            <Text className={styles.textBoxContent} maxLines={3} type="text1">
              Short info explanation about the feature will come here.
            </Text>
          </Flex>
        ),
        description: "Use text3 for section titles, short captions, and labels."
      },
      negative: {
        component: (
          <Text maxLines={5} type="text3" style={{ width: "70%" }}>
            monday CRM lets you control your entire sales funnel and close more deals, faster. Automate manual work and
            streamline sales activities from start to finish. Explore monday sales CRM templates
          </Text>
        ),
        description: "Don't use text3 as body text."
      }
    },
    {
      positive: {
        component: (
          <Flex direction="column" align="start" gap="xs" className={styles.textBoxContainer}>
            <Flex style={{ width: "90%" }} gap="small">
              <Icon icon={Link} />
              <Text color="secondary" maxLines={3} type="text3">
                Section title in text3
              </Text>
            </Flex>
            <Text className={styles.textBoxContent} maxLines={3} type="text1">
              Short info explanation about the feature will come here.
            </Text>
          </Flex>
        ),
        description: "Use only normal and medium weights for text3."
      },
      negative: {
        component: (
          <Text maxLines={5} weight="bold" type="text3" style={{ width: "70%" }}>
            monday CRM lets you control your entire sales funnel and close more deals, faster. Automate manual work and
            streamline sales activities from start to finish. Explore monday sales CRM templates
          </Text>
        ),
        description: "Don't use bold weight for text3, to ensure optimal readability."
      }
    }
  ]}
/>

## Related components

<RelatedComponents componentsNames={[HEADING, EDITABLE_HEADING, LINK]} />
