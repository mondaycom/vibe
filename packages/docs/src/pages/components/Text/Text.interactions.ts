import { type Screen } from "@testing-library/react";
import { interactionSuite } from "@vibe/core/interactionsTests";
import { testHoverTooltipTrigger } from "../Tooltip/Tooltip.interactions";

// Test constants
export const ONE_LINE_ELLIPSIS_TEST_ID = "ellipsis-text-example";
export const OVERFLOW_TEXT_CONTAINER_ID = "overflow-text-examples-container";

async function isTooltipAppearOnHover(canvas: Screen) {
  const getText = () => canvas.findByTestId(ONE_LINE_ELLIPSIS_TEST_ID);

  await testHoverTooltipTrigger(canvas, getText);
}

export const textOverflowSuite: ReturnType<typeof interactionSuite> = interactionSuite({
  tests: [isTooltipAppearOnHover]
});
