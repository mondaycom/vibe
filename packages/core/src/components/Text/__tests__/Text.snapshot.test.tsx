import { describe, it, expect } from "vitest";
import React from "react";
import renderer from "react-test-renderer";
import Text from "../Text";

describe("Text renders correctly", () => {
  it("with empty props", () => {
    const tree = renderer.create(<Text>text</Text>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("with type", () => {
    const tree = renderer.create(<Text type="text1">text</Text>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("with weight", () => {
    const tree = renderer.create(<Text weight="medium">text</Text>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("with color", () => {
    const tree = renderer.create(<Text color="secondary">text</Text>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("with element", () => {
    const tree = renderer.create(<Text element="p">text</Text>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("with align", () => {
    const tree = renderer.create(<Text align="center">text</Text>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("with element as children", () => {
    const tree = renderer
      .create(
        <Text>
          <div>text</div>
        </Text>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
