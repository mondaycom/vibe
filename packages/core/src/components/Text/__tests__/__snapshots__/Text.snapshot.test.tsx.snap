// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text renders correctly with align 1`] = `
<div
  className="typography primary center singleLineEllipsis text text2Normal"
  data-testid="text"
>
  text
</div>
`;

exports[`Text renders correctly with color 1`] = `
<div
  className="typography secondary start singleLineEllipsis text text2Normal"
  data-testid="text"
>
  text
</div>
`;

exports[`Text renders correctly with element 1`] = `
<p
  className="typography primary start text text2Normal"
  data-testid="text"
>
  text
</p>
`;

exports[`Text renders correctly with element as children 1`] = `
<div
  className="typography primary start singleLineEllipsis text text2Normal"
  data-testid="text"
>
  <div>
    text
  </div>
</div>
`;

exports[`Text renders correctly with empty props 1`] = `
<div
  className="typography primary start singleLineEllipsis text text2Normal"
  data-testid="text"
>
  text
</div>
`;

exports[`Text renders correctly with type 1`] = `
<div
  className="typography primary start singleLineEllipsis text text1Normal"
  data-testid="text"
>
  text
</div>
`;

exports[`Text renders correctly with weight 1`] = `
<div
  className="typography primary start singleLineEllipsis text text2Medium"
  data-testid="text"
>
  text
</div>
`;
